deep_quoridor/src/agents/alphazero/nn_evaluator.py

```diff
@@ -45,12 +45,16 @@ def evaluate(self, game: Quoridor):
         # Mask the policy to ignore invalid actions
         valid_actions = game.get_valid_actions()
         valid_action_indices = [self.action_encoder.action_to_index(action) for action in valid_actions]
-        policy_masked = np.zeros_like(unmasked_policy)
+        policy_masked = np.full_like(unmasked_policy, -1e32)
         policy_masked[valid_action_indices] = unmasked_policy[valid_action_indices]
 
-        # Re-normalize probabilities after masking.
-        policy_probs = policy_masked
-        policy_probs = policy_probs / policy_probs.sum()
+        if np.all(policy_masked == 0):
+            print("Policy is all zeros")
+            print(policy_masked)
+            print(game)
+
+        # Apply softmax to convert to probabilities
+        policy_probs = F.softmax(torch.from_numpy(policy_masked), dim=0).numpy()
```

Comment:
Hmm... if you take the softmax here, then you need to take the logarithm of the training data you use for the network, right? Otherwise the network is training on probabilities, but you're treating its output as log probabilities.

Reply:
:thinking:

Reply:
ok i think i may have found it! For the policy loss, they use  [tf.nn.softmax_cross_entropy_with_logits_v2(
            logits=policy_logits, labels=policy_targets), ](https://github.com/google-deepmind/open_spiel/blob/b11f9eb62a67a0982c0183d7170be6411c7447ab/open_spiel/python/algorithms/alpha_zero/model.py#L283), which knows that it is comparing log-probabilities with probabilities. We just use `cross_entropy()` loss, which expects both inputs to be probability distributions.